### PR TITLE
ElectricLoadCenter:Storage:LiIonNMCBattery

### PR DESCRIPTION
**Pull request by nmerket**
Pull request overview
---------------------
This PR adds a Li-ion battery option as a new object `ElectricLoadCenter:Storage:LiIonNMCBattery` in addition to the simple and kinetic battery models. This module uses the SAM Simulation Core (SSC) library for the battery simulation under the hood. 

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [x] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [x] If structural output changes, add to output rules file and add OutputChange label
 - [x] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: idd/Energy+.idd.in [hunk 1]
```diff
@@ -88754,6 +88754,128 @@ ElectricLoadCenter:Storage:Battery,
...
         \note Only required when battery life calculation is activated.
 
+ElectricLoadCenter:Storage:LiIonNMCBattery,
+        \memo Uses Lithium Ion NMC model to simulate rechargeable battery banks in an
+        \memo electrical load center. The battery bank is a collection of one or more individual
+        \memo battery modules. Given the surplus or deficit power from the electrical system and
+        \memo the state of charge from the previous time step, this object can model the voltage,
+        \memo current, and energy losses with charging and discharging during each time step.
+        \memo The cumulative battery damage can be also modeled and reported at the end of
+        \memo each simulation run.
```
Comment (mjwitte) [inline idd/Energy+.idd.in:88764]: For good measure, add `\min-fields 22`

File: doc/input-output-reference/src/overview/group-electric-load-center-generator.tex [hunk 2]
```diff
@@ -1406,6 +1406,105 @@ \subsubsection{Outputs}\label{electricloadcenter-storagebattery-outputs}
 
 This output reports the fractional battery life used up at a point of time. For example, a value of 0.4 at the end of one year simulation means that the 40\% of the battery life is used up, so the battery needs to be replaced every two and a half years.
 
+\subsection{ElectricLoadCenter:Storage:LiIonNMCBattery}\label{electricloadcenterstorage-liionbattery}
+
+This object usses the Li-ion NMC battery model from the System Advisor Model (SAM) software.
```
Comment (nmerket) [inline doc/input-output-reference/src/overview/group-electric-load-center-generator.tex:1411]: @rchintala13 Can you fill this in with more detail?

File: doc/input-output-reference/src/overview/group-electric-load-center-generator.tex [hunk 3]
```diff
@@ -1406,6 +1406,105 @@ \subsubsection{Outputs}\label{electricloadcenter-storagebattery-outputs}
...
+This field is included for those who want to modify the battery chemistry. Most users should leave this field blank or omit it to use the defaults, which are typical of Lithium-ion batteries.
+
+\paragraph{Field: Charge Rate at Which Voltage vs Capacity Curve Was Generated}\label{liion-crate}
+
+This field is included for those who want to modify the battery chemistry. Most users should leave this field blank or omit it to use the defaults, which are typical of Lithium-ion batteries.
+
+\paragraph{Field: Battery Cell Internal Electrical Resistance}\label{liion-internal-r}
+
+This field is included for those who want to modify the battery chemistry. Most users should leave this field blank or omit it to use the defaults, which are typical of Lithium-ion batteries.
+
```
Comment (nmerket) [inline doc/input-output-reference/src/overview/group-electric-load-center-generator.tex:1507]: Todo: IORef outputs.

File: idd/Energy+.idd.in (hunk 1 above)
Comment (nmerket) [reply idd/Energy+.idd.in:88764]: I modified this a bit since you saw it last and went with 11 as that's the minimum we need to run (and what most users will use).

File: src/EnergyPlus/ElectricPowerServiceManager.cc [hunk 4]
```diff
@@ -2943,11 +2945,14 @@ ElectricStorage::ElectricStorage( // main constructor
...
-      maxChargeRate_(0.0), lifeCalculation_(BatteyDegredationModelType::degredationNotSet), lifeCurveNum_(0), thisTimeStepStateOfCharge_(0.0),
-      lastTimeStepStateOfCharge_(0.0), pelNeedFromStorage_(0.0), pelFromStorage_(0.0), pelIntoStorage_(0.0), qdotConvZone_(0.0), qdotRadZone_(0.0),
-      timeElapsed_(0.0), thisTimeStepAvailable_(0.0), thisTimeStepBound_(0.0), lastTimeStepAvailable_(0.0), lastTimeStepBound_(0.0),
-      lastTwoTimeStepAvailable_(0.0), lastTwoTimeStepBound_(0.0), count0_(0), electEnergyinStorage_(0.0), thermLossRate_(0.0), thermLossEnergy_(0.0),
-      storageMode_(0), absoluteSOC_(0.0), fractionSOC_(0.0), batteryCurrent_(0.0), batteryVoltage_(0.0), batteryDamage_(0.0)
+      maxChargeRate_(0.0), lifeCalculation_(BatteryDegradationModelType::degredationNotSet), lifeCurveNum_(0),
+      liIon_dcToDcChargingEff_(0.0), liIon_mass_(0.0), liIon_surfaceArea_(0.0), liIon_Cp_(0.0), liIon_heatTransferCoef_(0.0), liIon_Vfull_(0.0),
+      liIon_Vexp_(0.0), liIon_Vnom_(0.0), liIon_Vnom_default_(0.0), liIon_Qfull_(0.0), liIon_Qexp_(0.0), liIon_Qnom_(0.0), liIon_C_rate_(0.0),
```
Comment (nmerket) [inline src/EnergyPlus/ElectricPowerServiceManager.cc:2950]: Added some class members to store the unique inputs for this model.

File: src/EnergyPlus/ElectricPowerServiceManager.cc [hunk 5]
```diff
@@ -3076,17 +3082,17 @@ ElectricStorage::ElectricStorage( // main constructor
             }
 
             if (UtilityRoutines::SameString(DataIPShortCuts::cAlphaArgs(6), "Yes")) {
-                lifeCalculation_ = BatteyDegredationModelType::lifeCalculationYes;
+                lifeCalculation_ = BatteryDegradationModelType::lifeCalculationYes;
```
Comment (nmerket) [inline src/EnergyPlus/ElectricPowerServiceManager.cc:3085]: Fixed an existing spelling error.

File: src/EnergyPlus/ElectricPowerServiceManager.cc [hunk 6]
```diff
@@ -3134,6 +3140,120 @@ ElectricStorage::ElectricStorage( // main constructor
             cutoffV_ = DataIPShortCuts::rNumericArgs(12);
             maxChargeRate_ = DataIPShortCuts::rNumericArgs(13);
 
+            break;
+        }
+        case StorageModelType::liIonNmcBattery: {
```
Comment (nmerket) [inline src/EnergyPlus/ElectricPowerServiceManager.cc:3145]: See below this for my part of the constructor.

File: src/EnergyPlus/ElectricPowerServiceManager.cc [hunk 7]
```diff
@@ -3134,6 +3140,120 @@ ElectricStorage::ElectricStorage( // main constructor
...
+                } else {
+                    // This sets a lifetime model where the capacity is always 100%.
+                    std::vector<double> tblVals{{20, 0, 100, 20, 5000, 100, 20, 10000, 100, 80, 0, 100, 80, 1000, 100, 80, 2000, 100}};
+                    util::matrix_t<double> battLifetimeMatrix(6, 3, &tblVals);
+                    battLifetime = new lifetime_calendar_cycle_t(battLifetimeMatrix, DataHVACGlobals::TimeStepSys);
+                }
+                std::vector<double> batt_losses{0};  // using double because SSC expects a double
+
+                // Create the SSC battery object
+                ssc_battery_ = std::unique_ptr<battery_t>(
```
Comment (nmerket) [inline src/EnergyPlus/ElectricPowerServiceManager.cc:3204]: Using the `battery_t` class from SAM Simulation Core (SSC) for the actual battery modeling.

File: src/EnergyPlus/ElectricPowerServiceManager.cc [hunk 8]
```diff
@@ -3191,6 +3304,9 @@ ElectricStorage::ElectricStorage( // main constructor
                 SetupEMSInternalVariable(state, "Electrical Storage Battery Maximum Capacity", name_, "[Ah]", maxAhCapacity_);
             }
         }
+        if (storageModelMode_ == StorageModelType::liIonNmcBattery) {
+            SetupOutputVariable(state, "Electric Storage Battery Temperature", OutputProcessor::Unit::C, batteryTemperature_, "System", "Average", name_);
+        }
```
Comment (nmerket) [inline src/EnergyPlus/ElectricPowerServiceManager.cc:3308]: This model has an internal temperature, so I included it as an output.

File: src/EnergyPlus/ElectricPowerServiceManager.cc [hunk 9]
```diff
@@ -3271,6 +3397,10 @@ void ElectricStorage::reinitAtBeginEnvironment()
             }
             batteryDamage_ = 0.0;
         }
+    } else if (storageModelMode_ == StorageModelType::liIonNmcBattery) {
+        // Copy the initial battery state to the last battery state
+        *ssc_lastBatteryState_ = *ssc_initBatteryState_;
+        ssc_battery_->set_state(*ssc_lastBatteryState_);
```
Comment (nmerket) [inline src/EnergyPlus/ElectricPowerServiceManager.cc:3402]: The battery model has this nifty state object that you can save and restore, which is really helpful for resetting after warmups and new environments and allowing us to run the same system timestep a few times as EnergyPlus does.

File: src/EnergyPlus/ElectricPowerServiceManager.cc [hunk 10]
```diff
@@ -3705,6 +3843,88 @@ void ElectricStorage::simulateKineticBatteryModel(EnergyPlusData &state,
...
 }
 
+void ElectricStorage::simulateLiIonNmcBatteryModel(EnergyPlusData &state,
+                                                   Real64 &powerCharge,
+                                                   Real64 &powerDischarge,
+                                                   bool &charging,
+                                                   bool &discharging,
+                                                   Real64 const controlSOCMaxFracLimit,
+                                                   Real64 const controlSOCMinFracLimit)
+{
```
Comment (nmerket) [inline src/EnergyPlus/ElectricPowerServiceManager.cc:3852]: Here's where we actually call the battery simulation.

File: src/EnergyPlus/ElectricPowerServiceManager.cc [hunk 11]
```diff
@@ -3725,6 +3945,17 @@ Real64 ElectricStorage::storedEnergy() const
...
+    return fractionSOC_;
+}
+
+Real64 ElectricStorage::batteryTemperature() const
+{
+    assert(storageModelMode_ == StorageModelType::liIonNmcBattery);
+    return batteryTemperature_;
+}
+
```
Comment (nmerket) [inline src/EnergyPlus/ElectricPowerServiceManager.cc:3957]: These are mostly helpful for unit testing.

File: testfiles/CMakeLists.txt [hunk 12]
```diff
@@ -499,6 +499,7 @@ add_simulation_test(IDF_FILE RetailPackagedTESCoil.idf EPW_FILE USA_IL_Chicago-O
 add_simulation_test(IDF_FILE RoomAirflowNetwork.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 add_simulation_test(IDF_FILE SeriesActiveBranch.idf EPW_FILE USA_IL_Chicago-OHare.Intl.AP.725300_TMY3.epw)
 add_simulation_test(IDF_FILE ShopWithPVandBattery.idf EPW_FILE USA_OK_Oklahoma.City-Will.Rogers.World.AP.723530_TMY3.epw)
+add_simulation_test(IDF_FILE ShopWithPVandLiIonBattery.idf EPW_FILE USA_OK_Oklahoma.City-Will.Rogers.World.AP.723530_TMY3.epw)
```
Comment (nmerket) [inline testfiles/CMakeLists.txt:504]: Created this new test file which is basically the same as ShopWithPBandBattery.idf, but with the Li-ion battery instead of the KiBaM battery.

**Review by nmerket — COMMENTED** [review]
Wanted to get others' review going on this. A lot of these files that are changing are updates to the SSC library. I have tried to mark the places where the integration happens. This is not quite complete. There are still a couple of pending things that I know of that will need to go in after the I/O freeze (and probably a few I don't):

- SSC is realeasing a version later this month with the battery model changes here in it. We'll need to update to that.
- I still owe you the rest of the IORef and Engineering Documentation.

**Comment by Myoldmopar** [comment]
I'm sure you saw it, but just for bookkeeping, @nmerket, there are build warnings and lots of failing tests in debug.

**Comment by nmerket** [comment]
@Myoldmopar Roger that. I saw the unit test failing about not having my internal gains wired up completely correctly. I suspect many of the errors are related to that. Looking into it now.

Regarding [the Custom Check failure](https://github.com/NREL/EnergyPlus/pull/8499/checks?check_run_id=1913733012), I'm not clear what the problem is. From what I can tell from the interface to that method, I'm supposed to set those variables depending on whether the battery can charge/discharge then. Can you explain to me what I'm supposed to be fixing there?

**Comment by Myoldmopar** [comment]
On the custom check failure, that particular check is in there because we had a number of cases in EnergyPlus where functions would accept a byref boolean variable, and inadvertently reset it to false before using it.  This was causing "ErrorsFound" to be reset to false in a number of worker functions around the code.  So many that it was justified to use a custom_check script to verify any functions that do this.  You can "fix" it by adding those to the allow list [here](https://github.com/NREL/EnergyPlus/blob/develop/scripts/dev/find_byref_bool_override.py#L109) if they are all acceptable.

File: src/EnergyPlus/DataHeatBalance.cc [hunk 13]
```diff
@@ -257,15 +257,15 @@ namespace DataHeatBalance {
...
     // Parameter for zone air flow mass balancing method
     int const AdjustMixingOnly(1);
     int const AdjustReturnOnly(2);
     int const AdjustMixingThenReturn(3);
     int const AdjustReturnThenMixing(4);
     int const NoAdjustReturnAndMixing(0);
+    // | 17 | 18 | 19 | 20 | 21 | 22 | 23 | 24 | 25 | 26 | 27 | 28 | 29 | 30 | 31 |
 
-    int const NumZoneIntGainDeviceTypes(53);
+    int const NumZoneIntGainDeviceTypes(54);
```
Comment (nmerket) [inline src/EnergyPlus/DataHeatBalance.cc:268]: Thanks for catching that. I thought I got all of those...

**Comment by Myoldmopar** [comment]
Well, that's certainly closer, but two new battery unit tests are segfaulting, it is causing an array out of bounds in one file, and the new battery example file is crashing.  I am not sure this will make it in right now.

**Comment by Myoldmopar** [comment]
There is definitely something weird going on here.  I am seeing the failed unit tests, and found that it is hitting a runtime error as such:

```
if (params->schedule_loss.empty()) {
        throw std::runtime_error("losses_t error: loss length must be greater than 0 for schedule mode");
    }
```

But it's weird, because that is called from the constructor, right after the `schedule_loss` variable is assigned a meaningful value.  Here is the calling sequence:

Inside `ElectricPowerServiceManager.cc`, the object is created:

```
new losses_t(batt_losses)
```

I don't want to talk about the use of `new` here.  I mean I do, but I realize that's a part of dealing with the ssc library apparently.  Anyway, the `losses_t` class has an overloaded constructor.  The one with a vector of double argument looks like this:

```
    explicit losses_t(const std::vector<double>& schedule_loss = std::vector<double>(1, 0)) {
        params = std::make_shared<losses_params>();
        params->loss_choice = losses_params::SCHEDULE;
        params->schedule_loss = schedule_loss;
        initialize();
    }
```

This is what I had to "fix" in f3c3b6b by moving the implementation into the header.  As you can see here, in line 3 of this constructor, the `schedule_loss` member variable on `params` is assigned to the passed in, or defaulted, function argument.  As I step through the debugger, I can verify this and everything is fine.  Then a call is made to `initialize()`, which is a method on the `losses_t` class.

As soon as the runtime switches to the new frame inside `initialize`, the instance variable appears to be empty again, and those assignments made in that constructor are gone.  The `params->schedule_loss` vector is empty, and that triggers the runtime exception.  

There are a number of things going on here.  There is a lot of manual and semi-manual memory management going on, but that's not necessarily a bad thing.  However, it can lead to weird ownership issues and stuff that I don't think we have time to debug.  The fact that we are calling a C++ library using `new` is very uncomfortable.  The fact that the `initialize()` function here, written in C++, has `goto` statements is mindblowing.

All that to say, after trying to debug it today, I am now even more confident this won't go in for 9.5.  AFAIK, this is the only thing holding up the freeze, so I am going to make an IO Freeze RC 1 right now, and let that build and test.  Then we'll chat tomorrow @nmerket about whether we think there is any way to remedy this tomorrow morning.  Either way, well have IO Freeze by tomorrow afternoon.

**Comment by Myoldmopar** [comment]
This now passes locally in a debug build.  I anticipate this merging soon once CI verifies it.

**Comment by Myoldmopar** [comment]
Fixed the unit test, but re-introduced the failures in the test files.  At least one is an array out of bounds exception, the other may be as well.  Also this reintroduced the windows build warnings in the unit tests where some variables were set to `float` instead of `Real64`.  They are unrelated to this work, but I had cleaned them up in the commit that was reverted.

**Comment by Myoldmopar** [comment]
Looking good so far @nmerket; Linux is pretty bogged down right now but I'll be watching.

**Comment by Myoldmopar** [comment]
Well, now there's some green!  Thanks @nmerket.
